File: sirius-core/src/main/java/org/starcoin/sirius/core/Update.java
package org.starcoin.sirius.core;

import org.apache.commons.lang3.RandomUtils;
import org.starcoin.core.*;
import org.starcoin.proto.Starcoin;
import org.starcoin.proto.Starcoin.ProtoUpdate;

import java.security.PrivateKey;
import java.security.PublicKey;
import java.util.List;
import java.util.Objects;

public class Update implements ProtobufCodec<ProtoUpdate>, Mockable {

    // transaction root hash.

    private int eon;
    // TODO need two version, one for data structure, another for content update.
    private long version;
    private long sendAmount;
    private long receiveAmount;

    private Hash root;

    private Signature sign;
    private Signature hubSign;

    public Update() {
    }

    //init by eon.
    public Update(int eon) {
        this.eon = eon;
        this.version = 0;
        this.sendAmount = 0;
        this.receiveAmount = 0;
        this.root = null;
    }

    public Update(int eon, long version, long sendAmount, long receiveAmount, Hash root) {
        this.eon = eon;
        this.version = version;
        this.sendAmount = sendAmount;
        this.receiveAmount = receiveAmount;
        this.root = root;
    }

    public Update(int eon, long version, BlockAddress address, List<OffchainTransaction> txs) {
        this.eon = eon;
        this.version = version;
        MerkleTree<OffchainTransaction> tree = new MerkleTree<>(txs);
        this.root = tree.hash();
        this.sendAmount =
                txs.stream()
                        .filter(transaction -> transaction.getFrom().equals(address))
                        .map(OffchainTransaction::getAmount)
                        .reduce(0L, Long::sum);
        this.receiveAmount =
                txs.stream()
                        .filter(transaction -> transaction.getTo().equals(address))
                        .map(OffchainTransaction::getAmount)
                        .reduce(0L, Long::sum);
    }

    public Update(ProtoUpdate update) {
        this.unmarshalProto(update);
    }

    public long getSendAmount() {
        return sendAmount;
    }

    public void setSendAmount(long sendAmount) {
        this.sendAmount = sendAmount;
    }

    public long getReceiveAmount() {
        return receiveAmount;
    }

    public void setReceiveAmount(long receiveAmount) {
        this.receiveAmount = receiveAmount;
    }

    public Hash getRoot() {
        return root;
    }

    public void setRoot(Hash root) {
        this.root = root;
    }

    public Signature getSign() {
        return sign;
    }

    public void setSign(Signature sign) {
        this.sign = sign;
    }

    public Signature getHubSign() {
        return hubSign;
    }

    public void setHubSign(Signature hubSign) {
        this.hubSign = hubSign;
    }

    public long getVersion() {
        return version;
    }

    public void setVersion(long version) {
        this.version = version;
    }

    public int getEon() {
        return eon;
    }

    public void setEon(int eon) {
        this.eon = eon;
    }

    public ProtoUpdate.Builder marshalSginData() {
        ProtoUpdate.Builder builder =
                ProtoUpdate.newBuilder()
                        .setEon(this.eon)
                        .setVersion(this.version)
                        .setSendAmount(this.sendAmount)
                        .setReceiveAmount(this.receiveAmount);
        if (this.root != null) {
            builder.setRoot(this.root.toProto());
        }
        return builder;
    }

    @Override
    public ProtoUpdate marshalProto() {
        ProtoUpdate.Builder builder = this.marshalSginData();
        if (this.sign != null) {
            builder.setSign(this.sign.toProto());
        }
        if (this.hubSign != null) {
            builder.setHubSign(this.hubSign.toProto());
        }
        return builder.build();
    }

    @Override
    public void unmarshalProto(ProtoUpdate proto) {
        this.eon = proto.getEon();
        this.version = proto.getVersion();
        this.sendAmount = proto.getSendAmount();
        this.receiveAmount = proto.getReceiveAmount();
        this.root = proto.hasRoot() ? Hash.wrap(proto.getRoot()) : null;
        this.sign = proto.hasSign() ? Signature.wrap(proto.getSign()) : null;
        this.hubSign = proto.hasHubSign() ? Signature.wrap(proto.getHubSign()) : null;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Update)) {
            return false;
        }
        Update update = (Update) o;
        return eon == update.eon
                && version == update.version
                && sendAmount == update.sendAmount
                && receiveAmount == update.receiveAmount
                && Objects.equals(root, update.root);
    }

    public static Update generateUpdate(Starcoin.ProtoUpdate protoUpdate) {
        if (protoUpdate == null) {
            return null;
        }
        Update update = new Update();
        update.unmarshalProto(protoUpdate);
        return update;
    }

    public boolean verifySig(PublicKey publicKey) {
        if (publicKey == null) {
            return false;
        }
        if (this.sign == null) {
            return false;
        }
        return this.sign.verify(publicKey, this.marshalSginData().build().toByteArray());
    }

    public boolean verifyHubSig(PublicKey publicKey) {
        if (publicKey == null) {
            return false;
        }
        if (this.hubSign == null) {
            return false;
        }
        return this.hubSign.verify(publicKey, this.marshalSginData().build().toByteArray());
    }

    public void sign(PrivateKey privateKey) {
        // TODO optimize resuse bytebuffer
        this.sign = Signature.of(privateKey, this.marshalSginData().build().toByteArray());
    }

    public void signHub(PrivateKey hubPrivateKey) {
        this.hubSign = Signature.of(hubPrivateKey, this.marshalSginData().build().toByteArray());
    }

    public boolean isSigned() {
        return this.sign != null;
    }

    public boolean isSignedByHub() {
        return this.hubSign != null;
    }

    @Override
    public int hashCode() {
        return Objects.hash(eon, version, sendAmount, receiveAmount, root);
    }

    @Override
    public void mock(MockContext context) {
        this.eon = RandomUtils.nextInt();
        this.version = RandomUtils.nextInt();
        this.sendAmount = RandomUtils.nextInt();
        this.receiveAmount = RandomUtils.nextInt();
        this.root = Hash.random();
    }
}

File: sirius-core/src/main/kotlin/org/starcoin/sirius/core/Update.kt
package org.starcoin.sirius.core

import kotlinx.serialization.SerialId
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import org.starcoin.proto.Starcoin
import org.starcoin.sirius.crypto.CryptoKey
import org.starcoin.sirius.crypto.CryptoService
import org.starcoin.sirius.serialization.BigIntegerSerializer
import org.starcoin.sirius.serialization.ProtobufSchema
import org.starcoin.sirius.serialization.toByteString
import org.starcoin.sirius.util.MockUtils
import java.math.BigInteger
import java.security.PrivateKey
import java.security.PublicKey


@ProtobufSchema(Starcoin.Update::class)
@Serializable
data class Update(
    @SerialId(1)
    val data: UpdateData,
    @SerialId(2)
    var sign: Signature = Signature.ZERO_SIGN,
    @SerialId(3)
    var hubSign: Signature = Signature.ZERO_SIGN
) : SiriusObject() {

    constructor() : this(0, 0L, 0L, 0L)

    constructor(
        eon: Int,
        version: Long,
        sendAmount: Long,
        receiveAmount: Long,
        root: Hash = Hash.EMPTY_DADA_HASH
    ) : this(
        UpdateData(
            eon,
            version,
            sendAmount.toBigInteger(),
            receiveAmount.toBigInteger(),
            root
        )
    )

    @JvmOverloads
    constructor(
        eon: Int,
        version: Long,
        sendAmount: BigInteger,
        receiveAmount: BigInteger,
        root: Hash = Hash.EMPTY_DADA_HASH
    ) : this(
        UpdateData(
            eon,
            version,
            sendAmount,
            receiveAmount,
            root
        )
    )

    @Transient
    val isSigned: Boolean
        get() = !this.sign.isZero()

    @Transient
    val isSignedByHub: Boolean
        get() = !this.hubSign.isZero()

    @Transient
    val eon: Int
        get() = data.eon

    @Transient
    val version: Long
        get() = data.version

    @Transient
    val sendAmount: BigInteger
        get() = data.sendAmount

    @Transient
    val receiveAmount: BigInteger
        get() = data.receiveAmount

    @Transient
    val root: Hash
        get() = data.root

    fun isEmpty(): Boolean {
        return this.data.isEmpty() && !this.isSigned && !this.isSignedByHub
    }

    fun verfySign(key: CryptoKey) = this.verifySig(key.keyPair.public)

    fun verifySig(publicKey: PublicKey): Boolean {
        return when {
            this.isSigned -> this.sign.verify(data, publicKey)
            else -> false
        }
    }

    fun verifyHubSig(key: CryptoKey) = this.verifyHubSig(key.keyPair.public)

    fun verifyHubSig(publicKey: PublicKey): Boolean {
        return when {
            this.isSignedByHub -> this.hubSign.verify(data, publicKey)
            else -> false
        }
    }

    fun sign(key: CryptoKey) {
        this.sign = key.sign(this.data)
    }

    fun sign(privateKey: PrivateKey) {
        this.sign(CryptoService.loadCryptoKey(privateKey))
    }

    fun signHub(key: CryptoKey) {
        this.hubSign = key.sign(this.data)
    }

    fun signHub(hubPrivateKey: PrivateKey) {
        this.sign(CryptoService.loadCryptoKey(hubPrivateKey))
    }

    companion object : SiriusObjectCompanion<Update, Starcoin.Update>(Update::class) {
        var DUMMY_UPDATE = Update(UpdateData.DUMMY_UPDATE_DATA)

        override fun mock(): Update {
            val update = Update(UpdateData.mock())
            val userKey = CryptoService.generateCryptoKey()
            val hubKey = CryptoService.generateCryptoKey()
            update.sign(userKey)
            update.sign(hubKey)
            return update
        }

        fun newUpdate(eon: Int, version: Long, address: Address, txs: List<OffchainTransaction>): Update {
            return Update(UpdateData.newUpdate(eon, version, address, txs))
        }
    }
}

@Serializable
@ProtobufSchema(Starcoin.UpdateData::class)
data class UpdateData(
    @SerialId(1)
    var eon: Int = 0,
    @SerialId(2)
    var version: Long = 0,
    @SerialId(3)
    @Serializable(with = BigIntegerSerializer::class)
    var sendAmount: BigInteger = BigInteger.ZERO,
    @SerialId(4)
    @Serializable(with = BigIntegerSerializer::class)
    var receiveAmount: BigInteger = BigInteger.ZERO,
    @SerialId(5)
    var root: Hash = Hash.EMPTY_DADA_HASH
) : SiriusObject() {

    constructor(
        eon: Int,
        version: Long,
        sendAmount: Long,
        receiveAmount: Long,
        root: Hash = Hash.EMPTY_DADA_HASH
    ) : this(eon, version, sendAmount.toBigInteger(), receiveAmount.toBigInteger(), root)

    fun isEmpty(): Boolean {
        return this.version == 0L && this.sendAmount == BigInteger.ZERO && this.receiveAmount == BigInteger.ZERO && root == Hash.EMPTY_DADA_HASH
    }

    companion object : SiriusObjectCompanion<UpdateData, Starcoin.UpdateData>(
        UpdateData::class
    ) {

        var DUMMY_UPDATE_DATA = UpdateData()

        override fun mock(): UpdateData {
            return UpdateData(
                MockUtils.nextInt(),
                MockUtils.nextLong(),
                MockUtils.nextBigInteger(),
                MockUtils.nextBigInteger(),
                Hash.random()
            )
        }

        fun newUpdate(eon: Int, version: Long, address: Address, txs: List<OffchainTransaction>): UpdateData {
            val tree = MerkleTree(txs)
            val root = tree.hash()
            val sendAmount = txs.stream()
                .filter { transaction -> transaction.from == address }
                .map<BigInteger> { it.amount }
                .reduce(BigInteger.ZERO) { a, b -> a.add(b) }
            val receiveAmount = txs.stream()
                .filter { transaction -> transaction.to == address }
                .map<BigInteger> { it.amount }
                .reduce(BigInteger.ZERO) { a, b -> a.add(b) }
            return UpdateData(eon, version, sendAmount, receiveAmount, root)
        }

        override fun parseFromProtoMessage(protoMessage: Starcoin.UpdateData): UpdateData {
            return UpdateData(
                protoMessage.eon,
                protoMessage.version,
                BigInteger(protoMessage.sendAmount.toByteArray()),
                BigInteger(protoMessage.receiveAmount.toByteArray()),
                Hash.wrap(protoMessage.root)
            )
        }

        override fun toProtoMessage(obj: UpdateData): Starcoin.UpdateData {
            return Starcoin.UpdateData.newBuilder()
                .setEon(obj.eon)
                .setVersion(obj.version)
                .setSendAmount(obj.sendAmount.toByteArray().toByteString())
                .setReceiveAmount(obj.receiveAmount.toByteArray().toByteString())
                .setRoot(obj.root.toByteString()).build()
        }
    }
}